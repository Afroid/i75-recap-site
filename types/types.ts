/**
 * Represents an individual content block within a recap.
 *
 * Each section can include structured content such as intro text, bullet points,
 * an outro, and an optional image (e.g., GIF or screenshot).
 */
export interface RecapSection {
  /** The name of the section (e.g., "intro", "tidbits", etc.) */
  type: string;

  /** Optional intro paragraph or lead-in for the section */
  intro?: string;

  /** Optional array of bullet points (supports Markdown formatting) */
  bullets?: string[];

  /** Optional outro paragraph or follow-up commentary */
  outro?: string;

  /** Optional fallback content if not using structured fields */
  content?: string;

  /** Optional image URL to include beneath the content (e.g., GIF, meme, screenshot) */
  imageUrl?: string;

  /** Alt text for the image (important for accessibility and SEO) */
  imageAlt?: string;
}

/**
 * Represents a single fantasy matchup with winner/loser info and analysis.
 */
export interface RecapMatchup {
  /** Team/player 1 name */
  team1: string;
  /** Team/player 2 name */
  team2: string;
  /** Name of the winner */
  winner: string;
  /** Final score in format '123.4-110.3' */
  score: string;
  /** Markdown content describing the matchup */
  breakdown: string;
  /** Optional image (e.g., gif or meme related to the matchup) */
  imageUrl?: string;
  /** Alt text for image */
  imageAlt?: string;
}

/**
 * Special structure for the Game Notes section.
 */
export interface GameNotesSection {
  type: "gameNotes";
  pointsLeader: {
    name: string;
    points: number;
    notes?: string;
  };
  dud: {
    name: string;
    points: number;
    notes?: string;
  };
  matchups: RecapMatchup[];
}

/**
 * Represents a single weekly recap for a season.
 *
 * Each recap includes a week number, a title (for display),
 * and a list of content sections that make up the recap body.
 */
export interface RecapWeek {
  /** The week number of the recap (e.g., 1 for Week 1) */
  week: number;

  /** A custom title for the recap (e.g., "Week 1 Recap" or something funnier) */
  title: string;

  /** An array of content sections that make up the full recap */
  sections: (RecapSection | GameNotesSection)[];
}

/**
 * Represents a full set of recaps for a given fantasy season.
 *
 * The structure groups all recaps for a specific year into a single file.
 * You can load a different RecapData object for each season.
 */
export interface RecapData {
  /** The season year these recaps belong to (e.g., 2024) */
  year: number;

  /** All recaps that occurred during the regular season for that year */
  recaps: RecapWeek[];
}

export type NameMapping = {
  aliases: string[];
  username: string;
};
