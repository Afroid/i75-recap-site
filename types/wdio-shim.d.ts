/// <reference types="webdriverio" />
/// <reference types="@wdio/globals/types" />
/// <reference types="@wdio/mocha-framework" />

declare namespace WebdriverIO {
  interface Browser {
    getByTestId(id: string): ChainablePromiseElement<Element>;
    getByText(text: string):   ChainablePromiseElement<Element>;
  }
}
