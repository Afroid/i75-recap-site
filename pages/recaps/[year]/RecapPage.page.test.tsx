import React from "react";
import { render, screen } from "@testing-library/react";
import fs, { PathOrFileDescriptor } from "fs";
import RecapPage, { getStaticPaths, getStaticProps } from "@/pages/recaps/[year]/[week].page";
import { RecapData } from "@/types/types";
import type { GetStaticPropsContext, GetStaticPathsContext } from "next";

// Stub RecapLayout so we can assert it was called with the right props
jest.mock(
  "@/components/RecapLayout",
  () => ({
    __esModule: true,
    default: ({ recap, year }: { recap: { week: number }; year: string }) => (
      <div
        data-testid="stubbed-recap-layout"
      >{`Stubbed RecapLayout â€“ year=${year}, week=${recap.week}`}</div>
    ),
  })
);

// Mock fs so getStatic[Whatever] can be driven from tests
jest.mock("fs");

describe("RecapPage ([year]/[week].page.tsx)", () => {
  it("renders fallback when no recap is passed", () => {
    render(<RecapPage recap={null} year="2024" />);
    expect(screen.getByTestId("recap-not-found"))
      .toHaveTextContent("Recap not found");
  });

  it("renders the stubbed RecapLayout with correct props", () => {
    const dummy = { week: 7, title: "ignored", sections: [] };
    render(<RecapPage recap={dummy} year="2023" />);
    const stub = screen.getByTestId("stubbed-recap-layout");
    expect(stub).toBeInTheDocument();
    expect(stub).toHaveTextContent("year=2023");
    expect(stub).toHaveTextContent("week=7");
  });

  it("matches snapshot when a recap is provided", () => {
    const dummy = { week: 1, title: "ignored", sections: [] };
    const { asFragment } = render(<RecapPage recap={dummy} year="2025" />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("getStaticPaths()", () => {
  beforeEach(() => {
    (fs.readdirSync as jest.Mock).mockReturnValue([
      "2021.json",
      "notes.txt",
      "2020.json",
    ]);

    // return different JSON payloads depending on which file is being read
    jest
      .spyOn(fs, "readFileSync")
      .mockImplementation(
        (
          filePath: PathOrFileDescriptor,
        ): string | Buffer => {
          const fileName = filePath.toString();

          if (fileName.endsWith("2021.json")) {
            return JSON.stringify({
              year: 2021,
              recaps: [{ week: 1, title: "A", sections: [] }],
            });
          }

          if (fileName.endsWith("2020.json")) {
            return JSON.stringify({
              year: 2020,
              recaps: [
                { week: 3, title: "B", sections: [] },
                { week: 4, title: "C", sections: [] },
              ],
            });
          }
          throw new Error(`Unexpected path: ${name}`);
          // return ""; // fallback
        }
      );
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("builds one path per week in each JSON file, and sets fallback=false", async () => {
    const result = await getStaticPaths({} as GetStaticPathsContext);

    expect(result).toEqual({
      paths: [
        { params: { year: "2021", week: "week-1" } },
        { params: { year: "2020", week: "week-3" } },
        { params: { year: "2020", week: "week-4" } },
      ],
      fallback: false,
    });
  });

  it("loads Week 1 from 2021.json", async () => {
    const result = await getStaticProps({ params: { year: "2021", week: "week-1" } });
    expect(result).toEqual(
      expect
      .objectContaining({ props: { year: "2021", recap: expect.objectContaining({ week: 1 }) } })
    );
  });

  it("loads Week 4 from 2020.json", async () => {
    const result = await getStaticProps({ params: { year: "2020", week: "week-4" } });
    expect(result).toEqual(
      expect
      .objectContaining({ props: { year: "2020", recap: expect.objectContaining({ week: 4 }) } })
    );
  });
});

describe("getStaticProps()", () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("returns the correct recap and year when the file exists", async () => {
    // prepare fs.readFileSync to return a single recap with week=5
    const fake: RecapData = {
      year: 2022,
      recaps: [
        { week: 5, title: "ignored", sections: [] },
        { week: 6, title: "nope", sections: [] },
      ],
    };
    jest
      .spyOn(fs, "readFileSync")
      .mockReturnValue(JSON.stringify(fake));

    // Build a minimal context object
    const ctx = {
      params: { year: "2022", week: "week-5" },
    } as GetStaticPropsContext;

    const result = await getStaticProps(ctx);

    expect(result).toEqual({
      props: {
        year: "2022",
        recap: { week: 5, title: "ignored", sections: [] },
      },
    });
  });

  it("returns notFound: true if the JSON read throws", async () => {
    jest.spyOn(fs, "readFileSync").mockImplementation(() => {
      throw new Error("file not found");
    });

    const ctx = {
      params: { year: "2040", week: "week-1" },
    } as GetStaticPropsContext;

    const result = await getStaticProps(ctx);
    expect(result).toEqual({ notFound: true });
  });
});
