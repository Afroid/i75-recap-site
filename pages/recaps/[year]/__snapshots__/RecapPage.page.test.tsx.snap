// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecapPage ([year]/[week].page.tsx) matches snapshot when a recap is provided 1`] = `
<DocumentFragment>
  <div
    data-testid="recap-page"
  >
    <div
      data-testid="stubbed-recap-layout"
    >
      Stubbed RecapLayout â€“ year=2025, week=1
    </div>
  </div>
</DocumentFragment>
`;
